ledger/ledger-api-tests/infrastructure/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/TimeoutParticipantTestContext.scala

```diff
@@ -0,0 +1,497 @@
+import com.daml.ledger.api.v1.active_contracts_service.GetActiveContractsRequest
+import com.daml.ledger.api.v1.admin.config_management_service.{
+  GetTimeModelResponse,
+  SetTimeModelRequest,
+  SetTimeModelResponse,
+  TimeModel,
+}
+import com.daml.ledger.api.v1.admin.package_management_service.{
+  PackageDetails,
+  UploadDarFileRequest,
+}
+import com.daml.ledger.api.v1.admin.participant_pruning_service.PruneResponse
+import com.daml.ledger.api.v1.admin.party_management_service.PartyDetails
+import com.daml.ledger.api.v1.command_completion_service.{
+  Checkpoint,
+  CompletionEndRequest,
+  CompletionEndResponse,
+  CompletionStreamRequest,
+  CompletionStreamResponse,
+}
+import com.daml.ledger.api.v1.command_service.{
+  SubmitAndWaitForTransactionIdResponse,
+  SubmitAndWaitForTransactionResponse,
+  SubmitAndWaitForTransactionTreeResponse,
+  SubmitAndWaitRequest,
+}
+import com.daml.ledger.api.v1.command_submission_service.SubmitRequest
+import com.daml.ledger.api.v1.commands.Command
+import com.daml.ledger.api.v1.completion.Completion
+import com.daml.ledger.api.v1.event.CreatedEvent
+import com.daml.ledger.api.v1.ledger_configuration_service.LedgerConfiguration
+import com.daml.ledger.api.v1.ledger_offset.LedgerOffset
+import com.daml.ledger.api.v1.package_service.{GetPackageResponse, PackageStatus}
+import com.daml.ledger.api.v1.transaction.{Transaction, TransactionTree}
+import com.daml.ledger.api.v1.transaction_service.{
+  GetTransactionByEventIdRequest,
+  GetTransactionByIdRequest,
+  GetTransactionsRequest,
+  GetTransactionsResponse,
+}
+import com.daml.ledger.api.v1.value.{Identifier, Value}
+import com.daml.ledger.client.binding.{Primitive, Template}
+import com.daml.lf.data.Ref.HexString
+import com.daml.timer.Delayed
+import com.google.protobuf.ByteString
+import io.grpc.health.v1.health.HealthCheckResponse
+import io.grpc.stub.StreamObserver
+
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{ExecutionContext, Future}
+
+class TimeoutParticipantTestContext(timeoutScaleFactor: Double, delegate: ParticipantTestContext)
+    extends ParticipantTestContext {
+
+  override val ledgerId: String = delegate.ledgerId
+  override val applicationId: String = delegate.applicationId
+  override val endpointId: String = delegate.endpointId
+  override private[participant] def services = delegate.services
+  override private[participant] implicit val ec: ExecutionContext = delegate.ec
+  override def ledgerEndpoint: Endpoint = delegate.ledgerEndpoint
+  override def features: Features = delegate.features
+  override def referenceOffset: LedgerOffset = delegate.referenceOffset
+  override def nextKeyId: () => String = delegate.nextKeyId
+  override def nextUserId: () => String = delegate.nextUserId
+  override def delayMechanism: DelayMechanism = delegate.delayMechanism
+
+  override def currentEnd(): Future[LedgerOffset] =
+    withTimeout("Get current end", delegate.currentEnd())
```

Comment:
I'm fine with the approach but just wondering - is there any benefit from using delegation other than keeping the boilerplate separate and making the timeouts more explicit?

Reply:
Yes, didn't want to do that because I want to force anyone that adds a new method, to consider the timeout.
If not, I'm afraid that with time, we would add methods without timeouts...and it was also easier to make sure I add it to all of them by needing to implement all the methods, because of the sheer size of the trait